packages/js/src/utils/find-npm-dependencies.ts

```diff
@@ -0,0 +1,203 @@
+    if (ignoreFileFilters.some((f) => f(fileData.file))) continue;
+
+    for (const dep of fileData.deps) {
+      const target = fileDataDepTarget(dep);
+
+      // If the node is external, then read package info from `data`.
+      const externalDep = projectGraph.externalNodes[target];
+      if (externalDep?.type === 'npm') {
+        npmDeps[externalDep.data.packageName] = externalDep.data.version;
+        continue;
+      }
+
+      // If node is internal, then try reading package info from `package.json` (which must exist for this to work).
+      const workspaceDep = projectGraph.nodes[target];
+      if (!workspaceDep) continue;
+      const cached = seenWorkspaceDeps[workspaceDep.name];
+      if (cached) {
+        npmDeps[cached.name] = cached.version;
+      } else {
+        const packageJson = readPackageJson(workspaceDep, workspaceRoot);
+        if (packageJson) {
+          // This is a workspace lib so we can't reliably read in a specific version since it depends on how the workspace is set up.
+          // ASSUMPTION: Most users will use '*' for workspace lib versions. Otherwise, they can manually update it.
+          npmDeps[packageJson.name] = '*';
+          seenWorkspaceDeps[workspaceDep.name] = {
+            name: packageJson.name,
+            version: '*',
+          };
+        }
+      }
+    }
+  }
+}
+
+function readPackageJson(
+  project: ProjectGraphProjectNode,
+  workspaceRoot: string
+): null | {
+  name: string;
+  dependencies?: Record<string, string>;
+  optionalDependencies?: Record<string, string>;
+  peerDependencies?: Record<string, string>;
+} {
+  const packageJsonPath = join(
+    workspaceRoot,
+    project.data.root,
+    'package.json'
+  );
+  if (fileExists(packageJsonPath)) return readJsonFile(packageJsonPath);
+  return null;
+}
+
+function collectPeerAndOptionalDependencies(
+  workspaceRoot: string,
+  sourceProject: ProjectGraphProjectNode,
+  projectGraph: ProjectGraph,
+  npmDeps: Record<string, string>
+): void {
+  const packageJson = readPackageJson(sourceProject, workspaceRoot);
+  if (packageJson?.peerDependencies) {
+    for (const pkg of Object.keys(packageJson.peerDependencies)) {
+      const externalNode = projectGraph.externalNodes[`npm:${pkg}`];
+      if (externalNode) {
+        npmDeps[pkg] = externalNode.data.version;
+      }
+    }
+  }
+  if (packageJson?.optionalDependencies) {
+    for (const pkg of Object.keys(packageJson.optionalDependencies)) {
+      const externalNode = projectGraph.externalNodes[`npm:${pkg}`];
+      if (externalNode) {
+        npmDeps[pkg] = externalNode.data.version;
+      }
+    }
+  }
+}
+
+function collectHelperDependencies(
+  workspaceRoot: string,
+  sourceProject: ProjectGraphProjectNode,
+  projectGraph: ProjectGraph,
+  npmDeps: Record<string, string>
+): void {
+  for (const target of Object.values(sourceProject.data.targets)) {
+    if (target.executor === '@nx/js:tsc' && target.options?.tsConfig) {
```

Comment:
Knowing which targets we look for can simplify this.

They might have an executor with `@nx/js:tsc` but they are using something else for the `build`.